### PR TITLE
Migrate segment decryption logic into segment loader

### PR DESCRIPTION
**Pull request by gesinger**
Resolve key URIs during playlist load (when segment URIs are resolved). Abort both key and segment requests during segment loader aborts. Reset bandwidth calculation on key timeouts, and send up any key request errors. Abort both key and segment requests on mediachange.

### CODE REVIEW

File: src/segment-loader.js [hunk 1]
```diff
@@ -287,28 +315,74 @@
...
+      }
+
+      if (!this.xhr_.segmentXhr && !this.xhr_.keyXhr) {
+        this.xhr_ = null;
+        this.processResponse_();
+      }
     },
 
     processResponse_: function() {
+      var segmentInfo, playlist, segment, mediaIndex, segIv;
```
Comment (dmlap) [inline src/segment-loader.js:360]: Not in love with the name `segIv` but I know this was pulled over from the old code.

File: src/segment-loader.js [hunk 2]
```diff
@@ -287,28 +315,74 @@
...
+        // this is an encrypted segment
+        // if the media sequence is greater than 2^32, the IV will be incorrect
+        // assuming 10s segments, that would be about 1300 years
+        segIv = segment.key.iv || new Uint32Array([0, 0, 0, mediaIndex + playlist.mediaSequence]);
+
+        // incrementally decrypt the segment
+        new videojs.Hls.Decrypter(segmentInfo.encryptedBytes,
+                                  segment.key.bytes,
+                                  segIv,
+                                  (function(err, bytes) {
```
Comment (dmlap) [inline src/segment-loader.js:379]: Is there actually an `err` case for this callback? Should we do anything about it?

File: src/videojs-hls.js [hunk 3]
```diff
@@ -602,9 +603,6 @@ videojs.HlsHandler.prototype.updateDuration = function(playlist) {
  * state suitable for switching to a different video.
  */
 videojs.HlsHandler.prototype.resetSrc_ = function() {
```
Comment (dmlap) [inline src/videojs-hls.js:605]: This whole function is on my hit-list for removal.

File: test/playlist-loader_test.js [hunk 4]
```diff
@@ -203,6 +203,56 @@
...
+                             '#EXTINF:2.833,\n' +
+                             'http://example.com/000001.ts\n' +
+                             '#EXT-X-ENDLIST\n');
+    equal(loader.media().segments[0].key.resolvedUri,
+          window.location.protocol + '//' +
+          window.location.host + '/video/playlist/keys/key.php',
+          'resolved multiple relative paths for key URI');
+  });
+
+  test('recognizes containing playlist domain-relative key URLs', function() {
```
Comment (dmlap) [inline test/playlist-loader_test.js]: Missing a word?

File: test/segment-loader_test.js [hunk 5]
```diff
@@ -8,17 +8,18 @@
...
+        mediaSequence: conf && conf.mediaSequence ? conf.mediaSequence : 0,
         discontinuityStarts: [],
         segments: [],
         endList: true
       },
       count = Math.floor(time / 10),
       remainder = time % 10,
-      i;
+      i,
+      isEncrypted = conf && conf.isEncrypted;
```
Comment (dmlap) [inline test/segment-loader_test.js:22]: If this gets much more complicated, we could use `videojs.mergeOptions()` to handle merging defaults+settings.

File: src/segment-loader.js (hunk 2 above)
Comment (gesinger) [inline src/segment-loader.js:379]: In the Decrypter code I only saw one use of done, and err is always set to null.

File: test/playlist-loader_test.js (hunk 4 above)
Comment (gesinger) [inline test/playlist-loader_test.js]: I think I just need to rename it entirely.

**Pull request merged**